File: app/src/main/java/net/tjado/usbgadget/RootTask.java
package net.tjado.usbgadget;

import android.os.AsyncTask;
import android.util.Pair;

public class RootTask extends AsyncTask<Void, Void, Pair> {
    OnRootTaskListener listener;
    String[] mCommands;

    public interface OnRootTaskListener {
        void OnRootTaskFinish(Pair response);
    }

    public RootTask(String command, OnRootTaskListener listener) {
        mCommands = new String[] {command};
        this.listener = listener;
    }

    public RootTask(String[] commands, OnRootTaskListener listener) {
        mCommands = commands;
        this.listener = listener;
    }

    @Override
    protected Pair doInBackground(Void... params) {
        return ExecuteAsRootUtil.execute(mCommands);
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
    }

    @Override
    protected void onPostExecute(Pair result) {
        if( listener != null ) {
            listener.OnRootTaskFinish(result);
        }
    }
}

File: app/src/main/java/net/tjado/usbgadget/RootTask.kt
package net.tjado.usbgadget

import androidx.core.util.Pair as APair

class RootTask : CoroutineTask<Void?, APair<*, *>> {
    private var listener: OnRootTaskListener?
    private var mCommands: Array<String>

    fun interface OnRootTaskListener {
        fun onRootTaskFinish(response: APair<*, *>?)
    }

    constructor(command: String, listener: OnRootTaskListener?) {
        mCommands = arrayOf(command)
        this.listener = listener
    }

    constructor(commands: Array<String>, listener: OnRootTaskListener?) {
        mCommands = commands
        this.listener = listener
    }

    override fun doInBackground(vararg params: Void?): APair<*, *> {
        return ExecuteAsRootUtil.execute(mCommands)
    }

    override fun onPostExecute(result: APair<*, *>) {
        if (listener != null) {
            listener!!.onRootTaskFinish(result)
        }
    }
}